### ISSUE TITLE
Builds should not fail on get/put steps with `volume disappeared from worker` errors all the time

### ISSUE THREAD

**Issue by andrewedstrom** [concourse/concourse#5300]
## Bug Report

(In hush house), we constantly (multiple times per day) see Concourse builds fail on `get` or `put` steps with errors like `volume <guid> disappeared from worker 'workers-worker-4`.  This is super disruptive.

## Expected Results

Builds should not fail on get or put steps with this error. Or if they _have_ to fail, there should be some workaround, like automatically retrying that step. I don't want to have to add `attempts` to every get or put step in all of my pipelines.

## Actual Results

Builds fail this way all the time and all we can do is retrigger them manually. If it's a job that takes a long time to run this can set us back an hour or more. It also leads to a lot of pull requests being left in the `pending` state (we use https://github.com/telia-oss/github-pr-resource for our pull request status checks)

## Version Info

* Concourse version: 6.0.0
* Deployment type (BOSH/Docker/binary): hush-house
* Infrastructure/IaaS: hush-house
* Did this used to work? It has always been a problem, but we do see it much more often on hush-house than wings. Often it seems to correspond to a worker pod getting reaped by kubernetes for using too many resources.

**Comment by pivotal-jamie-klassen** [comment]
FWIW I think this is a problem caused by https://github.com/concourse/concourse-chart, and our [particular configuration](https://github.com/concourse/hush-house/blob/master/deployments/with-creds/hush-house/values.yaml) which does not use ephemeral workers. The chart has a fairly disruptive default behaviour - there is an initContainer which [deletes all worker volumes](https://github.com/concourse/concourse-chart/blob/master/templates/worker-statefulset.yaml#L68).

I think there was a good reason for not using ephemeral workers, but I don't remember what it was. Maybe @cirocosta, who set up the environment originally, remembers.

**Comment by andrewedstrom** [comment]
cc @aramprice

**Comment by gowrisankar22** [comment]
@cirocosta same issue with me as well with helm chart. Which configuration we should enable to fix the problem?

**Comment by staylor-ds** [comment]
This pull request looks like it may address the issue: https://github.com/helm/charts/pull/17891

**Comment by kuldeeps88** [comment]
Concourse-worker are running as deployment in my set-up still getting this issue intermittently,

Concourse version: 6.4.0

`{"timestamp":"2020-08-13T17:14:36.259281383Z","level":"error","source":"atc","message":"atc.checker.check-step.image.init-image.failed-to-fetch-image-for-container","data":{"check":500382,"container":"630cfcc1-8fbf-4963-7f38-5641f6aa6069","error":"volume 'XXXX-XXXX-4677-4fcc-XXXX' disappeared from worker 'concourse-worker-77cc58d6f4-8pn9s'","session":"19.2.1.1","step-name":"cluster-db"}}
{"timestamp":"2020-08-13T17:14:36.259326124Z","level":"error","source":"atc","message":"atc.checker.check-step.image.init-image.create.failed-to-construct-resource","data":{"check":500382,"container":"a9ab4e6d-1261-41b7-6a09-1115e94334ca","error":"volume 'XXXX-XXXX-4677-4fcc-XXXX' disappeared from worker 'concourse-worker-77cc58d6f4-8pn9s'","session":"19.2.1.1.4","step-name":"cluster-db"}}`

**Comment by schmurfy** [comment]
my concourse web and workers are all running on dedicated VM and I also get this error regularly, I have yet to find a real trigger.

**Comment by kuldeeps88** [comment]
@schmurfy I made few changes to my set-up and it helped me to reduce the frequent failure for the reported issue.

**Step1**: If your workers are running as statefulset then deploy them as Kubernetes deployment
`$ kubectl -n concourse get deploy
NAME                   READY   UP-TO-DATE   AVAILABLE   AGE
concourse-postgresql   1/1     1            1           403d
concourse-web          3/3     3            3           403d
concourse-worker       7/7     7            7           116d
$ kubectl -n concourse get sts
NAME               READY   AGE
concourse-worker   0/0     403d
`
**Step2**: bump the version of the concourse to `image: concourse/concourse:6.4.0`

**Step3:** also increase the number of connections on the connected Postgres
`spec:
      containers:
      - args:
        - -c
        - max_connections=250`

**Step4:** some concourse config tunning as required by your project

`       - name: CONCOURSE_SWEEP_INTERVAL
          value: 30s
        - name: CONCOURSE_CONNECTION_DRAIN_TIMEOUT
          value: 1h
        - name: CONCOURSE_HEALTHCHECK_TIMEOUT
          value: 5s
        - name: CONCOURSE_EPHEMERAL
          value: "true"
        - name: CONCOURSE_BAGGAGECLAIM_DRIVER
          value: naive
        - name: CONCOURSE_BAGGAGECLAIM_BTRFS_BIN
          value: btrfs
        - name: CONCOURSE_BAGGAGECLAIM_MKFS_BIN
          value: mkfs.btrfs
        - name: CONCOURSE_VOLUME_SWEEPER_MAX_IN_FLIGHT
          value: "5"
        - name: CONCOURSE_CONTAINER_SWEEPER_MAX_IN_FLIGHT
          value: "5"`

**Comment by Anonymous-Coward** [comment]
We did switch from stateful set to deployment, didn't help much, as long as the database was kept in place - which is the whole point of a database, to survive application upgrades or repairs. What did help was to manually delete all records from the tables containers and volumes inside the database. No pipelines get harmed in the process, all pipelines blocked due to checks failing with messages about missing volumes or containers are unblocked.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.
